actionpack/lib/action_controller/metal/request_forgery_protection.rb

```diff
@@ -228,19 +230,23 @@ def verify_authenticity_token # :doc:
         mark_for_same_origin_verification!
 
         if !verified_request?
-          if logger && log_warning_on_csrf_failure
-            if valid_request_origin?
-              logger.warn "Can't verify CSRF token authenticity."
-            else
-              logger.warn "HTTP Origin header (#{request.origin}) didn't match request.base_url (#{request.base_url})"
-            end
+          if valid_request_origin?
+            @warning_message = "Can't verify CSRF token authenticity."
+          else
+            @warning_message = "HTTP Origin header (#{request.origin}) didn't match request.base_url (#{request.base_url})"
           end
+          logger.warn @warning_message if logger && log_warning_on_csrf_failure
+
           handle_unverified_request
```

Comment:
What do you think about moving the warning message to the forgery_protection_message?
```ruby
      def verify_authenticity_token # :doc:
        mark_for_same_origin_verification!

        if !verified_request?
          protection_strategy = forgery_protection_strategy.new(self)
          logger.warn protection_strategy.warning_message if logger && log_warning_on_csrf_failure
          protection_strategy.handle_unverified_request
        end
      end
```

Reply:
Hmm ok, that probably doesn't work, because they don't map 1-on-1.

Reply:
Or maybe:
```ruby
      class Exception
        def initialize(controller)
          @controller = controller
        end

        def warning_message
          if @controller.valid_request_origin?
            "Can't verify CSRF token authenticity."
          else
            "HTTP Origin header (#{@controller.request.origin}) didn't match request.base_url (#{@controller.request.base_url})"
        end

        def handle_unverified_request
          raise ActionController::InvalidAuthenticityToken, warning_message
        end
      end
 ```

Reply:
Hello Petrik

Thanks for the review and the proposals.

I tried your last proposal but it means I will need to duplicate warning message. Because warning message is used in logs and exception. Also I had to make `valid_request_origin?` public in order to use it because it is private at the moment. 

I think the existing proposal is good enough. What do you think @p8 ?

Thanks

Reply:
@benoittgt Thanks for trying the proposal. 🙇  I agree we don't want to make `valid_request_origin?` public.

My biggest objection to the current solution is storing the message in `@warning_message` so that it's available in `handle_unverified_request`. I think this adds some indirection that might be confusing.
For example calling `handle_unverified_request` outside  of`verify_authenticity_token` won't have the `@warning_message` set.

We could store it in a local var instead and pass that to `handle_unverified_request`.
But that also isn't an option as we need to change the signature of `handle_unverified_request` which is a public method.

What do you think about extracting the warning message to a method?
```ruby
def unverified_request_warning_message
  if valid_request_origin?
    "Can't verify CSRF token authenticity."
  else
    "HTTP Origin header (#{request.origin}) didn't match request.base_url (#{request.base_url})"
   end
end
```
And call that:
```ruby
logger.warn unverified_request_warning_message if logger && log_warning_on_csrf_failure
```
```ruby
protection_strategy.warning_message = unverified_request_warning_message if protection_strategy.is_a?(ProtectionMethods::Exception)
```

Reply:
This is nice suggestion. Love it. I add it! https://github.com/rails/rails/pull/41822/commits/5846df309778f0b46cbff94d9f3130f2fcc3b192